File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
    public int answer = -1;
    public boolean[] visited;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        adventure(k, dungeons, 0);

        return answer;
    }

    public void adventure(int k, int[][] dungeons, int play) {
        answer = Math.max(answer, play);

        for (int i = 0; i < dungeons.length; i++) {
            if(!visited[i] && k>=dungeons[i][0] ) {
                visited[i] = true;
                adventure(k - dungeons[i][1], dungeons, play + 1);
                visited[i] = false;
            }
        }

    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def dfs(k, cnt, dungeons, visited):
    max_cnt = cnt
    for i in range(len(dungeons)):
        if k >= dungeons[i][0] and visited[i] == 0:
            visited[i] = 1
            max_cnt = max(max_cnt, dfs(k - dungeons[i][1], cnt + 1, dungeons, visited))
            visited[i] = 0
    return max_cnt

def solution(k, dungeons):
    visited = [0] * len(dungeons)
    result = dfs(k, 0, dungeons, visited)
    return result